File: tests/algorithms/prime_sieve/prime_sieve.go
package main

import (
	"fmt"
	"math"
	"time"
)

func sieveOfEratosthenes(n int) []int {
	if n < 2 {
		return []int{}
	}
	
	isPrime := make([]bool, n+1)
	for i := range isPrime {
		isPrime[i] = true
	}
	isPrime[0], isPrime[1] = false, false
	
	limit := int(math.Sqrt(float64(n)))
	
	for i := 2; i <= limit; i++ {
		if isPrime[i] {
			for j := i * i; j <= n; j += i {
				isPrime[j] = false
			}
		}
	}
	
	var primes []int
	for i := 2; i <= n; i++ {
		if isPrime[i] {
			primes = append(primes, i)
		}
	}
	
	return primes
}

func main() {
	n := 100000
	
	fmt.Printf("Finding all primes up to %d...\n", n)
	start := time.Now()
	
	primes := sieveOfEratosthenes(n)
	
	duration := time.Since(start)
	
	fmt.Printf("Result: Found %d primes\n", len(primes))
	if len(primes) > 0 {
		fmt.Printf("Largest prime: %d\n", primes[len(primes)-1])
	}
	fmt.Printf("Execution time: %.6f seconds\n", duration.Seconds())
}

File: tests/algorithms/prime_sieve/prime_sieve.rs
use std::time::Instant;

fn sieve_of_eratosthenes(n: usize) -> Vec<usize> {
    if n < 2 {
        return Vec::new();
    }
    
    let mut is_prime = vec![true; n + 1];
    is_prime[0] = false;
    is_prime[1] = false;
    
    let limit = (n as f64).sqrt() as usize;
    
    for i in 2..=limit {
        if is_prime[i] {
            let mut j = i * i;
            while j <= n {
                is_prime[j] = false;
                j += i;
            }
        }
    }
    
    (2..=n).filter(|&i| is_prime[i]).collect()
}

fn main() {
    let n = 100_000;
    
    println!("Finding all primes up to {}...", n);
    let start = Instant::now();
    
    let primes = sieve_of_eratosthenes(n);
    
    let duration = start.elapsed();
    
    println!("Result: Found {} primes", primes.len());
    if let Some(&largest) = primes.last() {
        println!("Largest prime: {}", largest);
    }
    println!("Execution time: {:.6} seconds", duration.as_secs_f64());
}